File: 012/Erlang/problem012.erl
-module(problem012).
-export([problem012/1]).

-import(numbers, [factors/1, triangle_number/1]).

problem012(N) -> problem012(1, N).

problem012(I, N) ->
  case length(factors(triangle_number(I))) >= N of
    true -> triangle_number(I);
    false -> problem012(I + 1, N)
  end.

File: 012/Haskell/Problem012.hs
module Problem012 (problem012) where

import Data.List

import Numbers

problem012 :: Integer -> Integer
problem012 n = 
  case find enoughFactors $ map triangleNumber [1, 2..] of
    Just t -> t
    Nothing -> error "What happened?"
  where
    enoughFactors i = (length $ factors i) >= fromInteger n
